File: leetcode/SwapNodesInPairs/swapPairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode dummy(0);
        dummy.next = head;
        ListNode *prev = &dummy, *cur = head;
        while (cur && cur->next) {
            ListNode *next = cur->next->next;
            prev->next = cur->next;
            prev->next->next = cur;
            cur->next = next;
            prev = cur;
            cur = cur->next;
        }
        return dummy.next;
    }
};

File: leetcode/SwapNodesInPairs/swapPairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode swapPairs(ListNode head) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ListNode fake = new ListNode(-1), prev = fake, cur = head, next;
        fake.next = head;
        while (cur != null && cur.next != null) {
            next = cur.next;
            cur.next = next.next;
            prev.next = next;
            next.next = cur;
            prev = cur;
            cur = cur.next;
        }
        return fake.next;
    }
}